src/aat/java/uk/gov/hmcts/ccd/fta/steps/BackEndFunctionalTestScenarioPlayer.java

```diff
@@ -0,0 +1,270 @@
+                    if (pathVariable.equals("uid") && theUser.getUid() != null) {
+                        aRequest.pathParam(pathVariable, theUser.getUid());
+                        scenarioContext.getTestData().getRequest().getPathVariables().put("uid", theUser.getUid());
+                    } else {
+                        Assert.fail("Dynamic value for request path variable '" + pathVariable + "' does not exist");
+                    }
+                } else {
+                    aRequest.pathParam(pathVariable, value);
+                }
+            });
+        }
+
+        if (requestData.getQueryParams() != null) {
+            requestData.getQueryParams().forEach((queryParam, value) -> {
+                if (value.toString().equals(DYNAMIC_CONTENT_PLACEHOLDER)) {
+                    // ADD DYNAMIC DATA HERE
+                    Assert.fail("Dynamic value for request query parameter '" + queryParam + "' does not exist");
+                } else {
+                    aRequest.queryParam(queryParam, value);
+                }
+            });
+        }
+
+        if (requestData.getBody() != null) {
+            aRequest.body(new ObjectMapper().writeValueAsBytes(requestData.getBody()));
+        }
+
+        scenarioContext.setTheRequest(aRequest);
+        scenario.write(JsonUtils.getPrettyJsonFromObject(scenarioContext.getTestData().getRequest()));
+    }
+
+    @Override
+    @When("it is submitted to call the [{}] operation of [{}]")
+    public void submitTheRequestToCallAnOperationOfAProduct(String operation, String productName) throws IOException {
+        boolean isCorrectOperation = scenarioContext.getTestData().meetsOperationOfProduct(operation, productName);
+        String errorMessage = "Test data does not confirm it is calling the following operation of a product: "
+            + operation + " -> " + productName;
+        Assert.assertTrue(errorMessage, isCorrectOperation);
+
+        RequestSpecification theRequest = scenarioContext.getTheRequest();
+        String uri = scenarioContext.getTestData().getUri();
+
+        Response response = null;
+        Assert.assertNotNull("No request method in data file", scenarioContext.getTestData().getMethod());
+        switch (scenarioContext.getTestData().getMethod()) {
+            case "GET":
+                response = theRequest.get(uri);
+                break;
+            case "POST":
+                response = theRequest.post(uri);
+                break;
+            case "PUT":
+                response = theRequest.put(uri);
+                break;
+            case "DELETE":
+                response = theRequest.delete(uri);
+                break;
+            default:
+                Assert.fail("Unknown request method in data file");
+        }
+
+        QueryableRequestSpecification queryableRequest = SpecificationQuerier.query(theRequest);
+        scenario.write(queryableRequest.getMethod() + " " + queryableRequest.getURI());
```

Comment:
1. Ideally, the logging should happen before making the rest call.
2. We have created a QueryableRequestSpecification object for logging purpose. Can't we log the requestMethod and URI from the "theRequest" object?

Reply:
this is the earliest we can log the URI because the path variables are only populated once the call is made, and the RequestSpecification object needs to be made queryable otherwise the URI can't be retrieved